mautic/mautic
# Removes Mautic configuration's dependency on the container

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? | Yes and more Coming
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/8316
| BC breaks? | 
| Deprecations? | Yes

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR removes Mautic's dependency on Symfomy's container for its configuration options using Symfony 3's support for environment variables. This means that the Container no longer needs to be rebuilt when Mautic's configuration UI is updated. There are a few exceptions that require editing of the local config file such as API rate limiting that still requires rebuilding the cache. But the goal is that anything that can be edited through the configuration UI should not require a rebuild of the container.  

Unfortunately this is big and touches several major areas due to the need to refactor away from cache compiler passes. 

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Mautic should load :-D
3. Test that the Configuration saves and the values are persisted to app/config/local.php
4. Test sending emails both immediately and spooled to the filesystem
5. Test a SAML implementation (https://www.mautic.org/docs/en/authentication/saml.html; I personally tested using SF as an identify provider https://help.salesforce.com/articleView?id=identity_provider_enable.htm&type=5). If using Salesforce, configure the SAML Email attribute to be "Username", First name = "FirstName" and Last name = "LastName")
6. Test the QueueBundle (need help on the steps for this). The UI can no longer be used for this. You have to edit app/config/local.php manually. See https://github.com/mautic/mautic/pull/8265/commits/f6941e808bb4abd348034aad47dffb858c81d152
7. Test Mautic's installer
8. Test that API endpoints (including auth endpoints) are blocked if API is disabled. Also test that basic auth is rejected if disabled.
9. Test installing a new language through Configuration; log out and back in; the new language should be loaded
10. Test reloading plugins; clear out the plugins table then run `php bin/console -e prod mautic:plugins:reload`. Check /s/plugins that they are again listed.
11. Add `'debug' => true,` to app/config/local.php. Run the command to update the segment `php bin/console mautic:segments:update -e prod`. Check var/logs to see if there is a mautic_prod-YYYY-MM-DD.php file with debug info for building the segments. Now change debug to false and run the command again and no new DEBUG output should be written to the log. 
12. Test El Finder https://github.com/mautic/mautic/pull/8306
13. Send an email that contains links to contacts via CLI (campaigns are currently broken in 3.x so I had to use the `mautic:broadcast:send` command with a broadcast email that had a publish at and down date configured). The URLs should be correct based on what's configured as the `site_url` in app/config/local.php 

#### List backwards compatibility breaks:
1. See UPGRADE-3.0.md
